### PR TITLE
dealii 8.3. (update and fix)

### PR DESCRIPTION
**Pull request by davydden**
temporary requier trilinos to be built without python.

see https://github.com/Homebrew/homebrew-science/issues/2148

### CODE REVIEW

File: dealii.rb [hunk 1]
```diff
@@ -35,7 +35,11 @@ class Dealii < Formula
   depends_on "p4est"        => [:recommended] + openblasdep if build.with? "mpi"
   depends_on "slepc"        => :recommended
   # TODO: We won't have bottles if we make Trilinos :recommended because the bot times out.
-  depends_on "trilinos"     => [:optional] + openblasdep
+  # TODO: When trilinos 11.12.1 is built with Python, libpytrilinos is included through
+  # cmake configure files in deal.II. This leads to a run-time error:
+  # Symbol not found: _PyBool_Type
+  # see https://github.com/Homebrew/homebrew-science/issues/2148
+  depends_on "trilinos"     => [:optional] + openblasdep + ["without-python"]
```
Comment (dpo) [inline dealii.rb]: Could you please make that `["without-python", :optional] + openblasdep`?

**Comment by davydden** [comment]
fixed

**Comment by dpo** [comment]
Does the most recent Trilinos fix this?

**Comment by davydden** [comment]
i don't know, i have not tried compiling 12.0.1

**Comment by wfspotz** [comment]
Correct, it is not fixed yet.

**Comment by ashkan2200** [comment]
Why have you removed "without-python" from trillions installation in dealii formula? I just installed deal.ii and I have this problem. Should I do any tricks?

**Comment by davydden** [comment]
@ashkan2200 see https://github.com/Homebrew/homebrew-science/issues/2148 . This PR will fix it. I hope @dpo merge it soon

**Comment by dpo** [comment]
@BrewTestBot test this please

**Comment by davydden** [comment]
@dpo sorry to bother, but was the bot run successful?

**Comment by tseemann** [comment]
@BrewTestBot test this please.

**Comment by dpo** [comment]
Apologies. I've been extremely busy. I'll get to this very soon.

**Comment by dpo** [comment]
@BrewTestBot test this please

**Comment by dpo** [comment]
@BrewTestBot test this please

**Comment by davydden** [comment]
what's wrong with BrewTestBot, why do you have to invoke the run 4 time already?

p.s. there is 8.3. in release preparation. It should be out in a week or so. Since the whole PR is already around since end of May, we may just wait a bit and have an updated version.

**Comment by dpo** [comment]
Seems we'll need to pull this after PETSc anyways:
```
In file included from /usr/local/include/petscbag.h:4:
/usr/local/include/petscsys.h:130:6: error: "PETSc was configured with one OpenMPI mpi.h version but now appears to be compiling using a different OpenMPI mpi.h version"
#    error "PETSc was configured with one OpenMPI mpi.h version but now appears to be compiling using a different OpenMPI mpi.h version"
```

**Comment by davydden** [comment]
The scheduled release is 1 August.

**Comment by dpo** [comment]
@BrewTestBot test this please

**Comment by dpo** [comment]
Will this still be compatible with #2507?

**Comment by davydden** [comment]
sure. I am using it with `petsc 3.6`, `slepc 3.6`, trilinos `12.0.1`.

**Comment by davydden** [comment]
but please wait until I update to `8.3`, which should be around 1st of August.

**Comment by davydden** [comment]
i updated to 8.3., @dpo please give it a look.

**Comment by dpo** [comment]
```
[ 98%] Linking CXX shared library ../lib/libdeal_II.g.dylib
Undefined symbols for architecture x86_64:
  "_SuiteSparse_calloc", referenced from:
      _cholmod_l_calloc in libcholmod.a(cholmod_l_memory.o)
  "_SuiteSparse_config", referenced from:
      _umf_l_realloc in libumfpack.a(umf_l_realloc.o)
      _check_common in libcholmod.a(cholmod_l_check.o)
      _cholmod_l_gpu_stats in libcholmod.a(cholmod_l_check.o)
      _check_sparse in libcholmod.a(cholmod_l_check.o)
      _check_dense in libcholmod.a(cholmod_l_check.o)
      _check_subset in libcholmod.a(cholmod_l_check.o)
      _check_perm in libcholmod.a(cholmod_l_check.o)
      ...
  "_SuiteSparse_realloc", referenced from:
      _cholmod_l_realloc in libcholmod.a(cholmod_l_memory.o)
ld: symbol(s) not found for architecture x86_64
```

What version of SuiteSparse does this need? We now have `suite-sparse` (version 4.4.4) and `suite-sparse421` (version 4.2.1).

**Comment by davydden** [comment]
i have `4.4.4` locally and it works. I will investigate the issue next week, when i'm back from the trip.

**Comment by davydden** [comment]
i just tried installing it via homebrew, it did fine with `gcc 5.1` and no other related outdated formulae. 
how do you get that linking error? is it on test bot or on your local machine? is everything updated?

p.s. i also rebased it against the master so that it uses the latest homebrew formulae. Maybe that caused the issue.

**Comment by dpo** [comment]
The error message is copied from the test bot.

**Comment by davydden** [comment]
don't understand it as it links fine here... i will try updating gcc and see if it makes a difference...

**Comment by davydden** [comment]
@dpo could you pls run bot again? I rebased everything on top of master. Just to be sure that the linking error is still there and is the same?

**Comment by dpo** [comment]
Your updates triggered a build. Thanks for looking into this.

**Comment by davydden** [comment]
i see that build is finished, but i couldn't figure out what is the result

**Comment by dpo** [comment]
Bot error.

**Comment by dpo** [comment]
@BrewTestBot test this please

**Comment by dpo** [comment]
We're missing a dependency on `doxygen` now:
```
-- Could NOT find Doxygen (missing:  DOXYGEN_EXECUTABLE) 
CMake Error at doc/doxygen/CMakeLists.txt:32 (MESSAGE):
  Could not find doxygen which is required for building the documentation
```

**Comment by davydden** [comment]
my bad, tried to improve the formula but forgot to add doxygen dependecy. fixed it.

**Comment by davydden** [comment]
@BrewTestBot test this please

**Comment by davydden** [comment]
interesting...
```
Undefined symbols for architecture x86_64:
  "_SuiteSparse_calloc", referenced from:
      _cholmod_l_calloc in libcholmod.a(cholmod_l_memory.o)
  "_SuiteSparse_config", referenced from:
      _umf_l_realloc in libumfpack.a(umf_l_realloc.o)
      _check_common in libcholmod.a(cholmod_l_check.o)
      _cholmod_l_gpu_stats in libcholmod.a(cholmod_l_check.o)
      _check_sparse in libcholmod.a(cholmod_l_check.o)
      _check_dense in libcholmod.a(cholmod_l_check.o)
      _check_subset in libcholmod.a(cholmod_l_check.o)
      _check_perm in libcholmod.a(cholmod_l_check.o)
      ...
  "_SuiteSparse_realloc", referenced from:
      _cholmod_l_realloc in libcholmod.a(cholmod_l_memory.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [lib/libdeal_II.8.3.0.dylib] Error 1
make[1]: *** [source/CMakeFiles/deal_II.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```
@jppelteret Jean-Paul, did you hit the same error messages recently when trying to re-install deal.ii by hand?

**Comment by dpo** [comment]
That's a `suite-sparse` vs. `suite-sparse421` problem.

**Comment by davydden** [comment]
Are u daying that 'suite-sparse421' is/got installed on test bot and thus messes up dealii installation? 

I dont have '421' installed. But i assume it should be keg-only, so i dont quite see how would it mess up dealii installation

**Comment by dpo** [comment]
See https://github.com/Homebrew/homebrew-science/pull/1314#issuecomment-61303533.

Your `suite-sparse` might pre-date the time `suite-sparse421` was added to the repository. I'm not sure what's going on on the bot though.

**Comment by davydden** [comment]
That's surely not the case, as i have '4.4.4' installed. 
I have no outdated formulae on my machine, yosemite.

My guess would be that bottles of petsc were not updated after the switch to suite-sparse '4.4.4' in homebrew, thus the problem on the test-bot.

I propose you u to check the formula manually.

**Comment by davydden** [comment]
oh, and btw, `dealii` does not list `suite-sparse` as a dependency so it surely has to do with it's primary dependents. Among those which are build by default `petsc` (and therefore `slepc`) depend on `suite-sparse`.

**Comment by davydden** [comment]
i updated the formula (added some missing dependencies) and test it again in my environment **with** `suite-sparse421` (from octave) and `suite-sparse` installed. All good, the presence of `421` does not break anything, upstream `suiste-sparse` is used trhoughout :

```
$ brew install dealii --with-trilinos
...
==> Downloading https://github.com/dealii/dealii/releases/download/v8.3.0/dealii-8.3.0.tar.gz
Already downloaded: /Library/Caches/Homebrew/dealii-8.3.0.tar.gz
==> cmake .. -DCMAKE_BUILD_TYPE=DebugRelease -DCMAKE_INSTALL_PREFIX=/usr/local/Cellar/dealii/8.3.0 -DCMAKE_FIND_FRAMEWORK=LAST -Wno-dev -DDEAL_II_COMPONENT_COMPAT_FILES=OFF -DDEAL_II_COMPONENT_EXAMPLES=ON -DDEAL_II_COMPONENT_MESH_CONVERTER=ON -DCMAKE_PREFIX_PATH=/usr/loca
==> make
==> make test 2>&1 | tee make-test.log
==> 100% tests passed, 0 tests failed out of 8
==> make install
```

**Comment by davydden** [comment]
installs fine on `kubuntu` (`brew install dealii.rb --without-opencascade --without-muparser --with-openblas`)

**Comment by davydden** [comment]
@dpo we are good on this PR.

**Comment by dpo** [comment]
Let me pull Trilinos first, which I'm trying to build locally.

**Comment by davydden** [comment]
but unless we make in Trilinos `depends_on :python => :optional` i think we won't have bottles for trillions? 
If there will be bottles, i need to make `depends_on "trilinos" => :recommended` here...

p.s. only `dealii` currently uses trillions, i am all for trilinos bottles and `:recommended` dependency here.

File: dealii.rb [hunk 2]
```diff
@@ -24,17 +25,24 @@ class Dealii < Formula
...
   depends_on "slepc"        => :recommended
+  depends_on "suite-sparse" => [:recommended] + openblasdep
+  depends_on "tbb"          => :recommended
   # TODO: We won't have bottles if we make Trilinos :recommended because the bot times out.
-  depends_on "trilinos"     => [:optional] + openblasdep
+  # TODO: When trilinos 11.12.1 is built with Python, libpytrilinos is included through
+  # cmake configure files in deal.II. This leads to a run-time error:
+  # Symbol not found: _PyBool_Type
+  # see https://github.com/Homebrew/homebrew-science/issues/2148
+  depends_on "trilinos"     => ["without-python", :optional] + openblasdep
```
Comment (dpo) [inline dealii.rb]: Does this still apply to Trilinos 12.0.1?

File: dealii.rb [hunk 3]
```diff
@@ -24,17 +25,24 @@ class Dealii < Formula
...
-  depends_on "petsc"        => [:recommended] + openblasdep
   depends_on "p4est"        => [:recommended] + openblasdep if build.with? "mpi"
+  depends_on "parmetis"     => :recommended if build.with? "mpi"
+  depends_on "petsc"        => [:recommended] + openblasdep
   depends_on "slepc"        => :recommended
+  depends_on "suite-sparse" => [:recommended] + openblasdep
+  depends_on "tbb"          => :recommended
   # TODO: We won't have bottles if we make Trilinos :recommended because the bot times out.
-  depends_on "trilinos"     => [:optional] + openblasdep
+  # TODO: When trilinos 11.12.1 is built with Python, libpytrilinos is included through
```
Comment (dpo) [inline dealii.rb]: We have bottles now!

File: dealii.rb (hunk 2 above)
Comment (davydden) [inline dealii.rb]: it seems so: https://github.com/Homebrew/homebrew-science/pull/2285#issuecomment-110001784

**Comment by dpo** [comment]
Should `trilinos` be `:recommended` now?

**Comment by davydden** [comment]
i guess it depends. If you are fine with https://github.com/Homebrew/homebrew-science/pull/2712 then i should make `depends_on "trilinos"     => [:recommended] + openblasdep`. Whereas if you feel strongly about keeping `python` on by default in trilinos, i guess we'll keep it as it is now (`depends_on "trilinos"     => ["without-python", :optional] + openblasdep`) so that there are bottles. Or are you saying that there still will be bottles of dealii if we have `depends_on "trilinos"     => ["without-python", :required] + openblasdep`?

**Comment by dpo** [comment]
If the Trilinos folks haven't come up with a patch for the Python issue, I guess we should just make it optional in `trilinos`, but that's really too bad. On the upside, there will be bottles for both `trilinos` and `dealii`.

**Comment by dpo** [comment]
@BrewTestBot test this please

**Comment by davydden** [comment]
@BrewTestBot test this please

**Comment by davydden** [comment]
@BrewTestBot test this please

**Comment by davydden** [comment]
@BrewTestBot test this please

**Pull request closed without merging**